File: Mobiles/ios/iOS-MobileO11yDemo/Features/CarConnectionAction/Models/CarInfo.swift
//
//  CarInfo.swift
//  iOS-MobileO11yDemo
//
//  Created by Robert Magnusson on 04.06.24.
//

import Foundation

struct CarInfo {
    let vin: String
    let model: CarModel
    let color: CarColor
    let productionDate: Date
    let softwareVersion: CarSoftwareVersion
}

class CarInfoFactory {
    func getRandomCarInfo() -> CarInfo {
        return CarInfo(
            vin: UUID().uuidString,
            model: CarModel.allCases.randomElement() ?? .turboWombat,
            color: CarColor.allCases.randomElement() ?? .cosmicBlue,
            productionDate: ProductionDateFactory().getRandomDate(),
            softwareVersion: CarSoftwareVersionFactory().getRandomVersion()
        )
    }
}

class ProductionDateFactory {
    func getRandomDate() -> Date {
        var dateComponents = DateComponents()
        dateComponents.year = Int.random(in: 2011...2023)
        dateComponents.month = Int.random(in: 1...12)
        dateComponents.day = Int.random(in: 1...28)

        let calendar = Calendar.current
        return calendar.date(from: dateComponents)!
    }
}

File: Mobiles/flutter/lib/core/domain_layer/car/car_info.dart
import 'package:equatable/equatable.dart';
import 'package:flutter_mobile_o11y_demo/core/domain_layer/car/car_color.dart';
import 'package:flutter_mobile_o11y_demo/core/domain_layer/car/car_model.dart';
import 'package:flutter_mobile_o11y_demo/core/domain_layer/car/car_software_version.dart';
import 'package:flutter_mobile_o11y_demo/core/utils/random_element.dart';

class CarInfo extends Equatable {
  const CarInfo({
    required this.vin,
    required this.model,
    required this.color,
    required this.productionDate,
    required this.softwareVersion,
  });

  final String vin;
  final CarModel model;
  final CarColor color;
  final DateTime productionDate;
  final CarSoftwareVersion softwareVersion;

  @override
  List<Object?> get props => [
        vin,
        model,
        color,
        productionDate,
        softwareVersion,
      ];
}

class CarInfoFactory {
  CarInfoFactory({
    required this.carSoftwareVersionFactory,
    required this.randomVinFactory,
  });

  CarSoftwareVersionFactory carSoftwareVersionFactory;
  RandomVinFactory randomVinFactory;

  CarInfo getRandomCarInfo() {
    return CarInfo(
      vin: randomVinFactory.getRandomVin(),
      model: CarModel.values.getRandomElement(),
      color: CarColor.values.getRandomElement(),
      productionDate: RandomElementDateTimeX.getRandomDate(),
      softwareVersion: carSoftwareVersionFactory.getRandomVersion(),
    );
  }
}
